src/mcp/server.ts

```diff
@@ -70,13 +70,41 @@ export function createServer(backend: ServerBackend, runHeartbeat: boolean): Ser
       startHeartbeat(server);
     }
 
+    // Set up automatic progress notifications for timeout prevention
+    const progressToken = request.params._meta?.progressToken;
+    let progressTimeout: NodeJS.Timeout | null = null;
+
+    if (progressToken) {
+      let progressValue = 0;
+      const sendPeriodicProgress = async (): Promise<void> => {
+        try {
+          await extra.sendNotification({
+            method: 'notifications/progress',
+            params: {
+              progressToken,
+              progress: progressValue,
+            },
+          });
+          progressValue = (progressValue + 1) % 100;
+          progressTimeout = setTimeout(sendPeriodicProgress, 5000);
+        } catch (error) {
+          serverDebug('Progress notification failed:', error);
+        }
+      };
+      progressTimeout = setTimeout(sendPeriodicProgress, 5000);
```

Comment:
Avoid passing async methods into sync handers, remove async/await and do extra.sendNotification().then(...).catch(logUnhandledError); fire-and-forget style